[tts][ACR-1713] Add new TTS engine APIs for activated mode information (#1716)
Samsung/tizen-docs: docs/application/native/guides/text-input/tts-engine.md

# Text-to-speech Engine

Text-to-speech Engine (TTSE) is an engine to synthesize voice from input texts and to send the synthesized sound data to TTS framework. 

> TTSE developers can provide TTSE service users, to apply TTSE with functions that is necessary to operate the engine.

Reviewer: This sentence sounds unclear to me, It would be better if you could reframe it.
For example:
**_TTSE developers can provide TTSE service users with the functions that are necessary to operate the engine._**

Author: Thank you for your suggestion.
I changed that sentence with your suggestion. Please check and review it one more time.

> The main features of the TTSE API include:

Reviewer: ```suggestion
The main features of the TTSE API includes the following:
```

Author: Done

- Preparing the TTSE service for use

>   You can connect the TTSE service application. [Basic TTSE Processes](#basic) and [Prerequisites](#prerequisites) are defined.

Reviewer: ```suggestion
  You can connect the TTSE service application. [Basic TTSE processes](#basic) and [prerequisites](#prerequisites) are defined.
```

Author: Done

- TTSE Parameters

  You can [set information](#parameters), which includes language and state.

- TTSE Retrieval information

  You can [get information](#info_retrieval), which includes command count and audio type.

- TTSE Send results

>   You can send the results to the engine service user. [TTSE Send results](#send_result), is defined.

Reviewer: ```suggestion
  You can send the results to the engine service user. [TTSE send results](#send_result), is defined.
```

Author: I changed this part like the followings:
```suggestion
- Send results

  You can [send the results](#send_result) to the engine service user.
```

Please review this one more time.

<a name="prerequisites"></a>
## Prerequisites

> To enable your application to use the TTSE functionality:

Reviewer: ```suggestion
To enable your application to use the TTSE functionality, follow these steps:
```

Author: Done

1. To use the functions and data types of the TTSE API (in [mobile](../../api/mobile/latest/group__CAPI__UIX__TTSE__MODULE.html) and [wearable](../../api/wearable/latest/group__CAPI__UIX__TTSE__MODULE.html) applications), include the `<ttse.h>` header file in your application:

    ```
    #include <ttse.h>
    ```

2. Create a structure `ttse_request_callback_s`:

    ```
    ttse_request_callback_s callback = {0,};
    ```

> 3. Register callback functions using `ttse_main()` function (The registered callback functions will be invoked when the TTSE service users request the TTSE services).

Reviewer: ```suggestion
3. Register callback functions using `ttse_main()` function (The registered callback functions will be invoked when the TTSE service users request the TTSE services):
```

Author: Done

    ```
    ttse_main(argc, argv, &callback);
    ```

<a name="basic"></a>

> ## TTSE Basic Processes

Reviewer: ```suggestion
## TTSE basic processes
```

Author: Done

> Using TTSE, you can:

Reviewer: ```suggestion
Using TTSE, you can do the following:
```

Author: Done

1. Create a structure using the `ttse_request_callback_s` function.

2. Implement callback functions.

>     [Add Event Callbacks](#add_event_callbacks), sets pitch, load/unload voice, starts synthesizing and, so on by registered callback functions.

Reviewer: ```suggestion
    [Add event callbacks](#add_event_callbacks), set pitch, load/unload voice, start synthesizing, and so on by registered callback functions.
```

Author: Done

3. Register callback functions using `ttse_main()` function.

<a name="add_event_callbacks"></a>

> ## Add Event Callbacks

Reviewer: ```suggestion
## Add event callbacks
```

Author: Done

> To register and define event callbacks for the TTSE service application:

Reviewer: ```suggestion
To register and define event callbacks for the TTSE service application, follow these steps:
```

Author: Done

> 1. The TTSE developer must register the `initialize()`, `deinitialize()`, `get_info()`, `foreach_voices()`, `is_valid_voice()`, `set_pitch()`, `load_voice()`, `unload_voice()`, `start_synth()`, `cancel_synth()`, `check_app_agreed()`, `need_app_credential()` callbacks.

Reviewer: ```suggestion
1. The TTSE developer must register the `initialize()`, `deinitialize()`, `get_info()`, `foreach_voices()`, `is_valid_voice()`, `set_pitch()`, `load_voice()`, `unload_voice()`, `start_synth()`, `cancel_synth()`, `check_app_agreed()`, `need_app_credential()` callbacks:
```

Author: Done

        static int private_data_requested_cb(const char* key, char** data);
        static void activated_mode_changed_cb(int activated_mode);

        void
        int main(int argc, char* argv[])
        {
            ttse_request_callback_s callback = {
                /* Add the mandatory callbacks */
            };

            ttse_main(argc, argv, &callback);

            ttse_set_private_data_set_cb(private_data_set_cb);
            ttse_set_private_data_requested_cb(private_data_requested_cb);
            ttse_set_activated_mode_changed_cb(activated_mode_changed_cb);

            ...
        }
        ```

4. Define the optional callbacks:
    ```
    static int
    private_data_set_cb(const char* key, const char* data)
    {
        return 0;	/* private data save succeed */
    }

    static int
    private_data_requested_cb(const char* key, char** data)
    {
        /* found key */
        *data = strdup("data");
        return 0;	/* private data set succeed */
    }

    static void
    activated_mode_changed_cb(int activated_mode)
    {
        return ;	/* get activated modes changed succeed */
    }
    ```

<a name="parameters"></a>

> ## TTSE Parameters

Reviewer: ```suggestion
## TTSE parameters
```

Author: Done

You can set the following parameters about the TTSE:

> - Send the current state using the `ttse_send_error()` function and send the error to a engine service user.

Reviewer: ```suggestion
- Send the current state using the `ttse_send_error()` function and send the error to an engine service user.
```

Author: Done

    ```
    void
    send_error(void)
    {
        int ret = TTSE_ERROR_NONE;

        ret = ttse_send_error(TTSE_ERROR_NONE, "msg", NULL);
        if(TTSE_ERROR_NONE != ret)
            /* Error handling */
    }
    ```

<a name="info_retrieval"></a>

> ## TTSE Information Retrieval

Reviewer: ```suggestion
## TTSE information retrieval
```

Author: Done

You can get the following information about the TTSE:

> - Get the speed range using the `ttse_get_speed_range()` function.

Reviewer: ```suggestion
- Get the speed range using the `ttse_get_speed_range()` function:
```

Author: Done

    ```
    void
    get_speed_range(void)
    {
        int ret = TTSE_ERROR_NONE;
        int min_speed = 0;
        int normal_speed = 0;
        int max_speed = 0;

        ret = ttse_get_speed_range(&min_speed, &normal_speed, &max_speed);
        if(TTSE_ERROR_NONE != ret)
            /* Error handling */
    }
    ```

> - Get the pitch range using the `ttse_get_pitch_range()` function.

Reviewer: ```suggestion
- Get the pitch range using the `ttse_get_pitch_range()` function:
```

Author: Done

    ```
    void
    get_pitch_range(void)
    {
        int ret = TTSE_ERROR_NONE;
        int min_pitch = 0;
        int normal_pitch = 0;
        int max_pitch = 0;

        ret = ttse_get_pitch_range(&min_pitch, &normal_pitch, &max_pitch);
        if(TTSE_ERROR_NONE != ret)
            /* Error handling */
    }
    ```

> - Get the activated modes using the `ttse_get_activated_mode()` function. According to modes of TTS clients connected to TTS engine service, the output argument `activated_mode` of `ttse_get_activated_mode()` is a bit sequence as `#TTSE_MODE_MASK_DEFAULT | #TTSE_MODE_MASK_SCREEN_READER`.

Reviewer: ```suggestion
- Get the activated modes using the `ttse_get_activated_mode()` function. According to modes of TTS clients connected to TTS engine service, the output argument `activated_mode` of `ttse_get_activated_mode()` is a bit sequence as `#TTSE_MODE_MASK_DEFAULT | #TTSE_MODE_MASK_SCREEN_READER`:
```

Author: Done

    ```
    void
    get_activated_modes(void)
    {
        int ret = TTSE_ERROR_NONE;
        int activated_mode;

        ret = ttse_get_activated_mode(&activated_mode);
        if(TTSE_ERROR_NONE != ret)
            /* Error handling */
    }
    ```

<a name="send_result"></a>

> ## TTSE Send results

Reviewer: ```suggestion
## TTSE send results
```

Author: I changed this subject like the following:
```suggestion
## Send results
```

Please review this again. Thank you.

You can send the following result information about the TTSE:

> - Send the synthesized sound data to the engine service user using the `ttse_send_result()` function.

Reviewer: ```suggestion
- Send the synthesized sound data to the engine service user using the `ttse_send_result()` function:
```

Author: Done

    ```
    void
    send_result(void)
    {
        int ret = TTSE_ERROR_NONE;
        void* pcms;
        unsigned int pcm_size;
        int rate = 16000;

		...

        ret = ttse_send_result(TTSE_RESULT_EVENT_START, pcms, pcm_size, TTSE_AUDIO_TYPE_RAW_S16, rate, NULL);

        if(TTSE_ERROR_NONE != ret)
            /* Error handling */
    }
    ```

> ## Related Information

Reviewer: ```suggestion
## Related information
```

Author: Done
